File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/lineChart/LineChartOpDesc.java
package edu.uci.ics.texera.workflow.operators.visualization.lineChart;

import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import edu.uci.ics.amber.engine.operators.OpExecConfig;
import edu.uci.ics.texera.workflow.common.metadata.InputPort;
import edu.uci.ics.texera.workflow.common.metadata.OperatorGroupConstants;
import edu.uci.ics.texera.workflow.common.metadata.OperatorInfo;
import edu.uci.ics.texera.workflow.common.metadata.OutputPort;
import edu.uci.ics.texera.workflow.common.metadata.annotations.AutofillAttributeName;
import edu.uci.ics.texera.workflow.common.metadata.annotations.AutofillAttributeNameList;
import edu.uci.ics.texera.workflow.common.operators.OneToOneOpExecConfig;
import edu.uci.ics.texera.workflow.common.tuple.schema.Attribute;
import edu.uci.ics.texera.workflow.common.tuple.schema.Schema;
import edu.uci.ics.texera.workflow.common.tuple.schema.OperatorSchemaInfo;
import edu.uci.ics.texera.workflow.operators.visualization.VisualizationOperator;

import java.util.ArrayList;
import java.util.List;

import static java.util.Collections.singletonList;
import static scala.collection.JavaConverters.asScalaBuffer;

/**
 * LineChart is a visualization operator that can be used to get tuples for line chart.
 * It returns tuples with data name (String) and at least one number (Integer or Double).
 * @author Mingji Han, Xiaozhen Liu
 *
 */
public class LineChartOpDesc extends VisualizationOperator {
    @JsonProperty(value = "name column", required = true)
    @JsonPropertyDescription("column of name (for x-axis)")
    @AutofillAttributeName
    public String nameColumn;

    @JsonProperty(value = "data column(s)", required = true)
    @JsonPropertyDescription("column(s) of data (for y-axis)")
    @AutofillAttributeNameList
    public List<String> dataColumns;

    @JsonProperty(value = "chart style", required = true)
    public LineChartEnum lineChartEnum;

    @Override
    public String chartType() {
        return lineChartEnum.getChartStyle();
    }

    @Override
    public OpExecConfig operatorExecutor(OperatorSchemaInfo operatorSchemaInfo) {
        if (nameColumn == null) {
            throw new RuntimeException("line chart: name column is null");
        }
        if (dataColumns == null || dataColumns.isEmpty()) {
            throw new RuntimeException("line chart: data column is null or empty");
        }
        return new OneToOneOpExecConfig(operatorIdentifier(), worker -> new LineChartOpExec(this, operatorSchemaInfo));
    }

    @Override
    public OperatorInfo operatorInfo() {
        return new OperatorInfo(
                "Line Chart",
                "View the result in line chart",
                OperatorGroupConstants.VISUALIZATION_GROUP(),
                asScalaBuffer(singletonList(new InputPort("", false))).toList(),
                asScalaBuffer(singletonList(new OutputPort(""))).toList());
    }

    @Override
    public Schema getOutputSchema(Schema[] schemas) {
        List<Attribute> attributes = new ArrayList<>();
        attributes.add(schemas[0].getAttribute(nameColumn));
        for(String s : dataColumns) attributes.add(schemas[0].getAttribute(s));
        return Schema.newBuilder().add(attributes).build();
    }
}

File: core/amber/src/main/scala/edu/uci/ics/texera/workflow/operators/visualization/lineChart/LineChartOpDesc.scala
package edu.uci.ics.texera.workflow.operators.visualization.lineChart

import com.fasterxml.jackson.annotation.{JsonProperty, JsonPropertyDescription}
import com.kjetland.jackson.jsonSchema.annotations.JsonSchemaTitle
import edu.uci.ics.texera.workflow.common.metadata.{OperatorGroupConstants, OperatorInfo}
import edu.uci.ics.texera.workflow.common.operators.PythonOperatorDescriptor
import edu.uci.ics.texera.workflow.common.tuple.schema.{Attribute, AttributeType, Schema}
import edu.uci.ics.amber.engine.common.workflow.{InputPort, OutputPort}
import edu.uci.ics.texera.workflow.operators.visualization.{
  VisualizationConstants,
  VisualizationOperator
}

import java.util
import scala.jdk.CollectionConverters.ListHasAsScala

class LineChartOpDesc extends VisualizationOperator with PythonOperatorDescriptor {

  @JsonProperty(value = "title", required = false, defaultValue = "Line Chart")
  @JsonSchemaTitle("Plot Title")
  @JsonPropertyDescription("The value for the plot tile")
  var title: String = ""

  @JsonProperty(value = "yLabel", required = false, defaultValue = "Y Axis")
  @JsonSchemaTitle("Y Label")
  @JsonPropertyDescription("the label for y axis")
  var yLabel: String = ""

  @JsonProperty(value = "xLabel", required = false, defaultValue = "X Axis")
  @JsonSchemaTitle("X Label")
  @JsonPropertyDescription("the label for x axis")
  var xLabel: String = ""

  @JsonProperty(value = "lines", required = true)
  var lines: util.List[LineConfig] = _

  override def getOutputSchema(schemas: Array[Schema]): Schema = {
    Schema.builder().add(new Attribute("html-content", AttributeType.STRING)).build()
  }

  override def operatorInfo: OperatorInfo =
    OperatorInfo(
      "Line Chart",
      "View the result in line chart",
      OperatorGroupConstants.VISUALIZATION_GROUP,
      inputPorts = List(InputPort()),
      outputPorts = List(OutputPort())
    )

  def createPlotlyFigure(): String = {
    val linesPart = lines.asScala
      .map { lineConf =>
        val colorPart = if (lineConf.color != "") {
          s"line={'color':'${lineConf.color}'}, marker={'color':'${lineConf.color}'}, "
        } else {
          ""
        }

        val namePart = if (lineConf.name != "") {
          s"name='${lineConf.name}'"
        } else {
          s"name='${lineConf.yValue}'"
        }

        s"""fig.add_trace(go.Scatter(
            x=table['${lineConf.xValue}'],
            y=table['${lineConf.yValue}'],
            mode='${lineConf.mode.getModeInPlotly}',
            $colorPart
            $namePart
          ))"""
      }

    s"""
       |        fig = go.Figure()
       |        ${linesPart.mkString("\n        ")}
       |        fig.update_layout(title='$title',
       |                   xaxis_title='$xLabel',
       |                   yaxis_title='$yLabel')
       |""".stripMargin
  }

  override def generatePythonCode(): String = {
    val finalCode =
      s"""
         |from pytexera import *
         |
         |import plotly.express as px
         |import plotly.graph_objects as go
         |import plotly.io
         |import numpy as np
         |
         |class ProcessTableOperator(UDFTableOperator):
         |
         |    # Generate custom error message as html string
         |    def render_error(self, error_msg) -> str:
         |        return '''<h1>Line chart is not available.</h1>
         |                  <p>Reason is: {} </p>
         |               '''.format(error_msg)
         |
         |    @overrides
         |    def process_table(self, table: Table, port: int) -> Iterator[Optional[TableLike]]:
         |        if table.empty:
         |            yield {'html-content': self.render_error("input table is empty.")}
         |            return
         |        ${createPlotlyFigure()}
         |        # convert fig to html content
         |        html = plotly.io.to_html(fig, include_plotlyjs='cdn', auto_play=False)
         |        yield {'html-content': html}
         |""".stripMargin
    finalCode
  }

  // make the chart type to html visualization so it can be recognized by both backend and frontend.
  override def chartType(): String = VisualizationConstants.HTML_VIZ
}
